### ISSUE TITLE
Kivy Release Plan (2.2.0)

### ISSUE THREAD

**Issue by misl6** [kivy/kivy#8167]
Kivy Release Plan (2.2.0)
=============================

This is the release plan for the upcoming kivy release.

Pre release
============

* [x] Update macOS dependencies by updating the versions in [tools/build_macos_dependencies.sh](https://github.com/kivy/kivy/blob/master/tools/build_macos_dependencies.sh) if any is available.
* [x] Update Linux dependencies by updating the versions in [tools/build_linux_dependencies.sh](https://github.com/kivy/kivy/blob/master/tools/build_linux_dependencies.sh) if any is available.
* [x] Update kivy_deps.xxx
  * [x] Following the [guide](https://github.com/kivy/kivy-sdk-packager/blob/master/win/README.md) On a branch go through all the deps in [kivy-sdk-packager](https://github.com/kivy/kivy-sdk-packager/tree/master/win) and increase the versions to trigger a build if a new version of the dep is available.
  * [x] Add wheel support for the most recent python version 3.x for each [dependency](https://github.com/kivy/kivy-sdk-packager/tree/master/.github/workflows).
  * [x] Once they all build, merge to master, using the `[publish xxx win]` syntax and it'll upload to the [server](https://kivy.org/downloads/ci/win/deps/) and to pypi directly.
  * [x] If you bumped any of the major or minor versions, update the pinned versions of the dependencies in [pyproject.toml](https://github.com/kivy/kivy/blob/master/pyproject.toml) and [setup.cfg](https://github.com/kivy/kivy/blob/master/setup.cfg).
* [x] Update [setup.cfg](https://github.com/kivy/kivy/blob/master/setup.cfg)
  * [x] maximum Cython version.
  * [x] `python_versions` range.
  * [x] `python_requires` value.
  * [x] Add/remove Python versions in `setup.py` (`Programming Language :: Python :: x.y`).
* [x] Update [License file](https://github.com/kivy/kivy/blob/master/LICENSE) dates
* [ ] Correct “deprecated” directives and remove any that have been deprecated long enough.
* [x] Add wheel support for the most recent python version 3.x on all platforms.
* [x] Update all the files in doc/sources/installation.
* [x] Ensure latest build works everywhere (tests pass)
  * [x] GitHub actions tests.
  * [x] GitHub actions [scheduled nightly wheels](https://github.com/kivy/kivy/actions?query=event%3Aschedule) are generated.
  * [x] [Kivy.app](https://kivy.org/downloads/ci/osx/app/) is building.
  * [x] Check the daily ppa is building correctly, and add any new OS to be built (https://code.launchpad.net/~kivy-team/+recipe/kivy-daily).
* [x] Do pre-release development version (marked as `x.y.x.dev0`) tagging `(git tag -a x.y.zdev0 -m "Tag vx.y.zdev0")`
   * [x] Wheels (kivy, kivy-examples) and source should be uploaded automatically to pypi, check they are there for [all OS](https://pypi.org/project/Kivy/#history). Otherwise, get them from the CI and upload.
   * [x] Open PR for pre-release on [conda-forge](https://github.com/conda-forge/kivy-feedstock) to see if it builds on all platforms and fix the recipe as needed.
   * [ ] Check if https://www.piwheels.org/project/kivy/ is having issues building kivy on its latest OS. (or ask the piwheels maintainer to import our cross-compiled wheels)
   * [x] Check the [release page](https://github.com/kivy/kivy/releases), a draft release should have been generated. Make sure it contains all the artifacts, and then discard the pre-release draft.
* [x] Announce a feature-freeze date
* [x] Do a pre-release RC by changing version e.g. `x.y.z.dev0` to `x.y.zrc1` tagging `(git tag -a x.y.zrc1 -m "Tag vx.y.zrc1")`.
  * [x] Wheels (kivy, kivy-examples) and source should be uploaded automatically to pypi, check they are there for [all OS](https://pypi.org/project/Kivy/#history). Otherwise, get them from the CI and upload.
  * [x] Open PR for pre-release on [conda-forge](https://github.com/conda-forge/kivy-feedstock) to see if it builds on all platforms and fix the recipe as needed.
  * [ ] Check if https://www.piwheels.org/project/kivy/ is having issues building kivy on its latest OS. (or ask the piwheels maintainer to import our cross-compiled wheels)
  * [x] Check the [release page](https://github.com/kivy/kivy/releases), a draft release should have been generated. Make sure it contains all the artifacts, and then discard the pre-release draft.

Iterate as needed.

Final Release
=============

* [x] Create GH milestone for the release after this.
* [x] Check current release milestone issues, retag those left to the next version and close.
* [x] Prepare release notes. Previous release notes for reference: https://github.com/kivy/kivy/releases.
  * [x] Check that all PRs merged since the last release is listed under the current milestone and make sure their labels are correct.
  * [x] Use github-cli to dump the changelog from PRs of the current milestone and add it to the repo's changelog. See [changelog_parser](https://github.com/kivy/kivy/blob/master/kivy/tools/changelog_parser.py) for full instructions.
  * [x] Generate contributor list by running `git shortlog -s 1.10.1..HEAD > log`, remove duplicate lines listing the same person under different names and than do a regex search (`\n +[0-9]+\t+(.+)`) and replace (`, $1`) on the log to get the comma separated list of names.
  * [x] Clearly list any major changes, deprecations, or removals that will occur in the upcoming version.
  * [ ] Validate release notes with at least one native English speaker
* [x] Remove dev0/rc from the [version](https://github.com/kivy/kivy/blob/master/kivy/_version.py) and tag release (set `RELEASE` in `kivy/_version.py` to `True`, commit the result and tag `git tag -a x.y.z -m "Tag vx.y.z"` and push `git push kivy master --tags`). Wait for CI to complete building and uploading all the wheels and files to https://kivy.org/downloads/ci/ and [PyPi](https://pypi.org/project/Kivy/) and [PyPi-examples](https://pypi.org/project/Kivy-examples/).
  * [ ] If the wheels were not uploaded to PyPi manually download files from the server or CI and upload to pypi (https://pypi.org/project/Kivy/) with `twine upload *`.
* [x] Make the release directory and upload release files for all platforms to the kivy server under https://kivy.org/downloads/x.y.z/:
  ```sh
  ssh -p 2458 root@kivy.org
  cd /web/downloads
  mkdir x.y.z
  ```
  Copy all the CI generated file to that folder.
    * [x] Windows, macOS, linux and rpi wheels.
    * [x] macOS Kivy.app files.
    * [x] source distribution.
    * [x] kivy examples.
* [x] Update branches:
  * [x] On the [docs](https://github.com/kivy/kivy-website-docs/tree/docs-stable) repo, branch the `docs-stable` into `docs-x.y.z-1`, where `x.y.z-1` is the last released kivy version.
  * [x] Merge master into stable and wait for the stable docs to have built as follows:
    ```sh
    git fetch kivy
    git checkout --track kivy/stable
    # x.y.z-1 is the last Kivy release version (it's named stable currently)
    git branch stable-x.y.z-1
    git checkout -b stable2
    git merge master
    ```
    Fix any merge issues and commit the merge if there are conflicts. This lets you fix issues on the stable2 branch without having to nuke local stable if things go wrong.

    Now make sure people on stable have a clean upgrade path:
    ```sh
    git checkout stable
    git checkout -b stable3
    git merge stable2
    ```
    If the stable2 merge was clean, you're ready to push stable with
    ```
    git push kivy stable-x.y.z-1:stable-x.y.z-1
    git push kivy stable2:docs_tmp
    ```
    Now make a PR to merge `docs_tmp` into `stable` (e.g. https://github.com/kivy/kivy/pull/6349), get it approved and merged.
  * [x] Delete `docs_tmp`.
  * [x] Wait for the stable docs to fully build on the CI and check that it appears on https://github.com/kivy/kivy-website-docs.
  * [x] Check that the kivy docs now list `stable` and `stable-x.y.z-1` and the contents contain what is expected.
* [x] Add the release announcement to the GitHub draft release on https://github.com/kivy/kivy/releases and publish it.
* [x] Make new conda release on https://github.com/conda-forge/kivy-feedstock. If you wait a little after the github tag/release, conda will make a PR to https://github.com/conda-forge/kivy-feedstock/pulls with the new release that you can just merge.
  * [x] Verify that the new release was [uploaded](https://anaconda.org/conda-forge/kivy).
* [x] Check that the [piwheels index](https://www.piwheels.org/project/kivy/) successfully compiled Kivy for the RPi. (or ask the piwheels maintainer to import our cross-compiled wheels)
* [x] Update the PPA (daily and stable, https://github.com/kivy/kivy-sdk-packager/tree/master/linux/debian).
  * [x] update rules/control/changelog files if changes required for daily and stable.
  * [x] Copy all changes from the daily control and rules file into the stable.
  * [x] update changelog file for daily and stable version in kivy-sdk-packager
  * [x] Edit the recipe (https://code.launchpad.net/~kivy-team/+recipe/kivy-stable) to build on the desired Ubuntu platforms - compare to https://code.launchpad.net/~kivy-team/+recipe/kivy-daily/ which already have been updated to master before the release.
  * [x] Request to re-import kivy-sdk-packager and kivy (https://code.launchpad.net/~thopiekar/kivy/+git/kivy-sdk-packager, https://code.launchpad.net/~thopiekar/kivy/+git/kivy).
  * [x] Wait to make sure that both kivy and kivy-sdk-packager has been imported on launchpad and that the stable branch lists the proper last commit from the merge from master.
  * [x] Trigger a new build for daily and stable.
  * [x] Wait to make sure daily and stable builds have properly uploaded.
* [x] Bump kivy version to the next version.
  * [x] Set `RELEASE` to `False` and `rcx` to `dev0`.
* [x] Website
  * [x] Update the announcements header
  * [ ] Update gallery with new projects
  * [x] Merge the website
  * [x] Ensure the page is updated
* [x] Update Discord announcement channel
* [ ] Share on Kivy Planet
* [x] Share on Google groups (kivy-dev + kivy-users)
* [ ] Share on Reddit (https://www.reddit.com/r/Python/comments/69uctq/kivy_1100_released/)
* [ ]  Share on HN (https://news.ycombinator.com/newest (only if it's a significant release, don’t direct link))
* [x] Share on Facebook
* [x] Add new announcement on [Launchpad](https://launchpad.net/kivy/+announcements).
* [x] Share on Twitter https://twitter.com/mathieuvirbel/status/861353662299484160

Methods
----------

* You can add `[build wheel]` to the commit message to force a build for wheels. Similarly `[build app osx]` to build the osx kivy.app files. You can also do `[build wheel win]`, `[build wheel linux]`, or `[build wheel osx]` to selectively build wheels for a platform.

Notes
--------

- Only `Ubuntu` versions with a proper `SDL` version will be supported via `ppa` (`Kinetic` and `Lunar`)
- Opened an issue on `piwheels` to manage the upload of our cross-compiled artifacts, as `piwheels` will be unable to build our dependencies (See: https://github.com/piwheels/packages/issues/356) - **UPDATE:** We're manually sending the artifacts to `piwheels` maintainer.

Related PRs:
- https://github.com/conda-forge/kivy-feedstock/pull/50

About deprecations: https://github.com/kivy/kivy/issues/8177

**Comment by quasistoic** [comment]
I see "Announce feature freeze" in the todo list. Does this mean I have a little time if I want to take a crack at fixing #8166.

**Comment by misl6** [comment]
> I see "Announce feature freeze" in the todo list. Does this mean I have a little time if I want to take a crack at fixing #8166.

We do not have announced it yet, since we have a few PRs left tagged with `2.2.0` milestone, so feel free to submit your PR.
If it will not make it for `2.2.0`, it will be surely merged after the release.

**Comment by misl6** [comment]
> I'm from Sri Lanka using kivy. I have developed a large app using kivy. I have a big problem about the rendering of sinhala font in kivy. Does kivy support sinhala font in this version?
> 
> As we want to support Sinhala languages ​​in our applications.

Please do not spam on this issue. I will reply shortly on #8148

**Comment by mablue** [comment]
I need this thing to use now bro I cant wait :)) pleaseeeeeeeeeeee

**Comment by misl6** [comment]
🟢  Release notes (to copy-paste): 🟢 


The Kivy team is happy to announce the release of Kivy 2.2.0.

Kivy is a full-featured framework for creating novel and performant user
interfaces, such as multi-touch applications, released under the **MIT license**.
The framework works on Windows, macOS, Linux, Android, iOS and Raspberry Pi.

Installation instructions and Kivy Docs are hosted on the [Kivy Website](https://www.kivy.org)

Currently supported Python versions are **3.7-3.11**

Highlights
----------

- [:repo:`7876`]: `Line`/`SmoothLine`: Fixes rendering issues related to corner radius and updates its order (`rounded_rectangle`) + add getter methods for `rounded_rectangle`, `rectangle`, `ellipse`, `circle`.
- [:repo:`7882`]: Re-implements the Bubble widget.
- [:repo:`7908`]: Speed up SmoothLine creation by ~2.5x
- [:repo:`7942`]: Config unicode support on Windows
- [:repo:`7988`]: Added support for KIVY_LOG_MODE
- [:repo:`8044`]: Add support for Python 3.11
- [:repo:`8056`]: New Feature: Add `BoxShadow` graphic instruction 🎉
- [:repo:`8115`]: Use `font_direction` and `font_script_name` from SDL2_ttf
- [:repo:`8144`]: Added property for mouse draggable tab scrollbar to TabbedPanel
- [:repo:`8162`]: `Label`: allow different values of left, top, right and bottom for `padding`.
- [:repo:`8169`]: `Image`: add `fit_mode` feature
- [:repo:`8096`]: Introduce build script for SDL dependencies and `KIVY_DEPS_ROOT`

Deprecated
----------

- [:repo:`7882`]: Re-implements the Bubble widget.

Breaking changes
----------------

- [:repo:`7876`]: `Line`/`SmoothLine`: Fixes rendering issues related to corner radius and updates its order (`rounded_rectangle`) + add getter methods for `rounded_rectangle`, `rectangle`, `ellipse`, `circle`.

Kv-lang
-------

- [:repo:`8021`]: Update builder.py

Misc
----

- [:repo:`7906`]: Replace deprecated logging.warn with logging.warning
- [:repo:`7913`]: fix(UrlRequest): Add "on_finish" and add alternative implementation
- [:repo:`7943`]: Fixes some E275 - assert is a keyword. + other minor PEP8 fixes
- [:repo:`7969`]: Config is not available when generating docs + Use `getdefault` instead of `has_option` + `get`

Widgets
-------

- [:repo:`7626`]: New Feature: Allow control how many lines to scroll at once using the mouse wheel on TextInput
- [:repo:`7882`]: Re-implements the Bubble widget.
- [:repo:`7905`]: Fix TextInputCutCopyPaste widget
- [:repo:`7925`]: Qwerty VKeyboard button fix( z, Q and W and ] ) on Linux(Ubuntu Focal Fossa)
- [:repo:`8109`]: Fix for changes of Splitter.strip_cls having no effect
- [:repo:`8144`]: Added property for mouse draggable tab scrollbar to TabbedPanel
- [:repo:`8169`]: `Image`: add `fit_mode` feature
- [:repo:`8202`]: Migrate `allow_stretch` and `keep_ratio` in widgets/examples by corresponding `fit_mode`

Core-app
--------

- [:repo:`7942`]: Config unicode support on Windows
- [:repo:`7958`]: Use AddLevelName in kivy.Logger to define TRACE
- [:repo:`7962`]: Refactored logging.ColoredFormatter to avoid deepcopy.
- [:repo:`7971`]: Support KivyLogMode environment variable for logging testing
- [:repo:`7973`]: Bump KIVY_CONFIG_VERSION and add a warning for future changes.
- [:repo:`7975`]: Light clean up of stderr handling code.
- [:repo:`7979`]: #7978: Don't monkey-patch logging.root
- [:repo:`7985`]: Handle non-strings in logs.
- [:repo:`7988`]: Added support for KIVY_LOG_MODE
- [:repo:`7989`]: Android Lifecycle convergence
- [:repo:`7994`]: Use urlopen instead of build_opener when fetching files from 'internet'. Removes some PY2 compat.
- [:repo:`8062`]: Use `find_spec`, `module_from_spec` and `exec_module` instead of `find_module` and `load_module` since are deprecated.

Core-providers
--------------

- [:repo:`7846`]: Fix VKeyboard missing with custom keyboard class
- [:repo:`7857`]: iOS camera provider enhancements
- [:repo:`7982`]: Use `SDL_WINDOWEVENT_DISPLAY_CHANGED` to notice about window switching display to update `_density` an `dpi`
- [:repo:`7999`]: Modify layout fix bug in how long text without space is cut 
- [:repo:`8025`]: Release the GIL when performing SDL_GL_SwapWindow call.
- [:repo:`8058`]: Makes Windows DPI aware of scale changes
- [:repo:`8076`]: New Feature: Always On Top
- [:repo:`8083`]: Allow changing `Window.fullscreen` and `Window.borderless` options after setup on iOS
- [:repo:`8115`]: Use `font_direction` and `font_script_name` from SDL2_ttf
- [:repo:`8142`]: New Feature: Allows to hide the taskbar icon
- [:repo:`8146`]: Fix memory issue on iOS 16.2 for AVMetadataObject (during QRCode scan)
- [:repo:`8147`]: Detect High DPI on Linux Desktop
- [:repo:`8162`]: `Label`: allow different values of left, top, right and bottom for `padding`.
- [:repo:`8171`]: Make VideoFFPy work with RTSP streams.
- [:repo:`8184`]: Revert "Detect High DPI on Linux Desktop"

Core-widget
-----------

- [:repo:`8035`]: Simplify Animation._unregister

Distribution
------------

- [:repo:`7837`]: Bump to 2.2.0.dev0
- [:repo:`7852`]: Build python 3.9 wheels for RPi
- [:repo:`7974`]: Bump SDL2, SDL_image, SDL_mixer, SDL_ttf versions to latest stable release
- [:repo:`8004`]: Bump kivy_deps.sdl2 and kivy_deps.sdl2_dev to 0.5.0
- [:repo:`8006`]: Use Platypus 5.4
- [:repo:`8043`]: Bump SDL2 to `2.24.1` on Linux and macOS
- [:repo:`8044`]: Add support for Python 3.11
- [:repo:`8050`]: Bump again SDL2 to 2.24.2 on Linux and macOS
- [:repo:`8070`]: Remove usage of `distutils` module which is deprecated and slated for removal in 3.12
- [:repo:`8096`]: Introduce build script for SDL dependencies and `KIVY_DEPS_ROOT`
- [:repo:`8155`]: Dependencies build tool: exit immediately on fail and allows to debug easier
- [:repo:`8173`]: Bump macOS dependencies versions on `tools/build_macos_dependencies.sh`
- [:repo:`8174`]: Bump Linux dependencies versions on `tools/build_linux_dependencies.sh`
- [:repo:`8176`]: Bump Windows dependencies via `kivy_deps` packages
- [:repo:`8178`]: Bump `cython_max` version
- [:repo:`8191`]: XCode 14.3 fails to build SDL if `MACOSX_DEPLOYMENT_TARGET` < `10.13`
- [:repo:`8203`]: Migrate from `autotools` to `cmake` for SDL2 linux dependencies
- [:repo:`8223`]: Perform RPi builds on `balenalib/raspberrypi3-*` images and skip `DISPMANX` API usage if can't be used [build wheel armv7l]
- [:repo:`8231`]: Bump version to `2.2.0rc1`

Documentation
-------------

- [:repo:`7870`]: Documentation: bump Gentoo install instructions
- [:repo:`7916`]: Fixes NO DOCUMENTATION (module kivy.uix.recycleview)
- [:repo:`7927`]: Fix minor typo in pong tutorial code comments
- [:repo:`7928`]: Add missing closing paren in hint text
- [:repo:`7929`]: Use consistent source code notes in pong tutorial
- [:repo:`7930`]: Purge trailing whitespace in docs source files
- [:repo:`7946`]: Add doc for `Canvas.add()`
- [:repo:`8026`]: Typo : missing coma in the doc
- [:repo:`8032`]: doc: Initial remarks on BSD compatibility.
- [:repo:`8034`]: Fix backticks typo in pong tutorial
- [:repo:`8039`]: Link to buildozer installation instructions instead of duplicating them
- [:repo:`8041`]: installation-osx.rst: Minor code formatting
- [:repo:`8088`]: Add support for sphinx `6.0.0`
- [:repo:`8089`]: Add a warning about `keyboard_suggestions` usage on Android
- [:repo:`8139`]: Improve docs about `BoxShadow` behavior and usage.
- [:repo:`8156`]: Docs: Update the Ubuntu prerequisites to build Kivy and its dependencies
- [:repo:`8175`]: Update Copyright and LICENSE dates
- [:repo:`8179`]: Update Python supported versions
- [:repo:`8181`]: :book: Grammar tweaks to focus docstrings
- [:repo:`8183`]: Docs: Fixes a typo (issue #7838)
- [:repo:`8229`]: Sphinx `7.0.0` is incompatible, use `<=6.2.1` for now
- [:repo:`8234`]: Docs review  for `RPi` installation and build instructions

Graphics
--------

- [:repo:`7860`]: Ellipse: update angle_start, angle_end to explicit floats
- [:repo:`7876`]: `Line`/`SmoothLine`: Fixes rendering issues related to corner radius and updates its order (`rounded_rectangle`) + add getter methods for `rounded_rectangle`, `rectangle`, `ellipse`, `circle`.
- [:repo:`7908`]: Speed up SmoothLine creation by ~2.5x
- [:repo:`8056`]: New Feature: Add `BoxShadow` graphic instruction 🎉
- [:repo:`8098`]: Fix `BoxShadow` shader crashing issue on Adreno GPUs
- [:repo:`8132`]: `BoxShadow`: Add `inset` feature
- [:repo:`8138`]: `BoxShadow`: Accept values for vertical and horizontal `spread_radius`
- [:repo:`8163`]: `Line`/`SmoothLine`: `ellipse` - fix behavior and add feature to allow closing line through center of ellipse
- [:repo:`8164`]: `Ellipse`: Handle the number of segments and avoid division by zero
- [:repo:`8170`]: Add svg rotation transform support
- [:repo:`8187`]: `Line`/`SmoothLine` - `ellipse`: Handle the number of segments to match `Ellipse`

Input
-----

- [:repo:`8027`]: Typo : German Keyboard is QWERTZ

Tests/ci
--------

- [:repo:`7847`]: Tests: ffpyplayer now ships cp310-* and Apple Silicon compatible wheels, so tests on the full version can be re-introduced.
- [:repo:`7854`]: Fixes 3.8.x pyenv install due to a recent change in clang [build wheel osx]
- [:repo:`7885`]: Our self-hosted Apple Silicon runner now has been migrated to actions/runner v2.292.0 which now supports arm64 natively
- [:repo:`7903`]: Migrate from probot/no-response to lee-dohm/no-response
- [:repo:`7917`]: When using pytest_asyncio for tests, function should be decorated with `pytest_asyncio.fixture`
- [:repo:`7972`]: Fix trivial typo in workflow.
- [:repo:`7987`]: Fix source typo in test_uix_bubbles.py
- [:repo:`8084`]: Switch from `ubuntu-18.04` to `ubuntu-latest` as `18.04` runners will be removed on 2023-01-12
- [:repo:`8093`]: Add `gstreamer1.0-plugins-good` for `autoaudiosink` availability during tests
- [:repo:`8099`]: Install twine only when needed [build wheel]
- [:repo:`8117`]: Upgrade GitHub Actions
- [:repo:`8120`]: [build wheel] Upgrade more GitHub Actions
- [:repo:`8121`]: GitHub Actions: Use current Python instead of hardcoded v3.9
- [:repo:`8126`]: Switch back to `macos-latest` instead of `macos-11`
- [:repo:`8129`]: Remove remaining nosetest settings in favor of pytest
- [:repo:`8157`]: Correct the flake8 pre-commit URL
- [:repo:`8217`]: `Generate-sdist` needs `packaging` as a dependency [build wheel win]


Thanks
--------

A big thanks to all of the Contributors, especially those github usernames mentioned here:

Akshay Arora, Alsp, Anders, AnthonyZimmermann, CdrSonan, Cecil Curry, Chris Vincent, Christian Clauss, Danny Dawson, Dexer, Evgeniy Selezniov, Gabriel Pettier, Gabriel Vogel, Hugo van Kemenade, Johan Toft, Julian, Matt Einhorn, Mirko Galimberti, NomadDemon, Paul Cochrane, Peter Tripp, Richard Larkin, Robert Niederreiter, RobertF, Robin Picard, Romain Heller, SanAndreyas, Tom Dudley, Tomasz CEDRO, Vinayak Sachan, anand vibhuti, djogy, luzpaz, mp-007, 水戸う納豆齋(Nattōsai Mitō)
